File: app/Models/IrrigationEvent.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class IrrigationEvent extends Model
{
    protected $fillable = [
        'greenhouse_id',
        'initiated_by',
        'type',
        'status',
        'started_at',
        'ended_at',
        'duration_minutes',
        'humidity_before',
        'humidity_after',
        'source',
        'notes',
    ];

    protected function casts(): array
    {
        return [
            'started_at' => 'datetime',
            'ended_at' => 'datetime',
            'humidity_before' => 'decimal:2',
            'humidity_after' => 'decimal:2',
        ];
    }

    public function greenhouse(): BelongsTo
    {
        return $this->belongsTo(Greenhouse::class);
    }

    public function initiator(): BelongsTo
    {
        return $this->belongsTo(User::class, 'initiated_by');
    }
}

File: lumatek_mobile/lib/features/irrigation/domain/irrigation_event.dart
import '../../../shared/models/json_utils.dart';

/// Evento de riego. Refleja App\Models\IrrigationEvent.
class IrrigationEvent {
  const IrrigationEvent({
    required this.id,
    required this.greenhouseId,
    this.type,
    this.status,
    this.startedAt,
    this.endedAt,
    this.durationMinutes,
    this.humidityBefore,
    this.humidityAfter,
    this.source,
    this.notes,
  });

  final int id;
  final int greenhouseId;

  /// manual | automatic | scheduled
  final String? type;

  /// running | completed | cancelled
  final String? status;

  final DateTime? startedAt;
  final DateTime? endedAt;
  final int? durationMinutes;
  final double? humidityBefore;
  final double? humidityAfter;
  final String? source;
  final String? notes;

  bool get isRunning => status == 'running';

  /// Cuánto lleva corriendo, para la cuenta regresiva del panel de riego.
  Duration? get elapsed {
    if (startedAt == null) return null;
    final end = endedAt ?? DateTime.now();
    return end.difference(startedAt!);
  }

  /// Tiempo restante estimado según la duración solicitada.
  Duration? get remaining {
    if (!isRunning || startedAt == null || durationMinutes == null) return null;
    final total = Duration(minutes: durationMinutes!);
    final passed = DateTime.now().difference(startedAt!);
    final left = total - passed;
    return left.isNegative ? Duration.zero : left;
  }

  double? get humidityGain {
    if (humidityBefore == null || humidityAfter == null) return null;
    return humidityAfter! - humidityBefore!;
  }

  factory IrrigationEvent.fromJson(Map<String, dynamic> json) {
    return IrrigationEvent(
      id: asInt(json['id']),
      greenhouseId: asInt(json['greenhouse_id']),
      type: asStringOrNull(json['type']),
      status: asStringOrNull(json['status']),
      startedAt: asDate(json['started_at']),
      endedAt: asDate(json['ended_at']),
      durationMinutes: asIntOrNull(json['duration_minutes']),
      humidityBefore: asDoubleOrNull(json['humidity_before']),
      humidityAfter: asDoubleOrNull(json['humidity_after']),
      source: asStringOrNull(json['source']),
      notes: asStringOrNull(json['notes']),
    );
  }
}
